### PR TITLE
Implement logger to workflow_engine and improve log messages

### PR DESCRIPTION
**Pull request by QU3B1M**
# Description

This PR implements a new Logger for the `workflow_engine` module. This logger is completely separated from the generic's one, letting the module to be used without installing any other module.

The default log level is `INFO`, anyway in that level, the logs raised from any other module/script wont be automatically logged in the console, to do so we must use the `DEBUG` level. That was defined that way in order to avoid collapsing the log file and the console itself with unwanted logs every time a workflow is executed.  

Also, the logs have different colors depending on its color, the color profile is the following:
```yaml
  DEBUG:    cyan
  INFO:     green
  WARNING:  yellow
  ERROR:    red
  CRITICAL: red,bg_white
```

---

## Testing performed

- **Execution with log level `INFO`**
    ```shellsession
    python workflow_engine.py /home/quebim/Wazuh/repos/wazuh-qa/deployability/modules/workflow_engine/examples/dtt1-agents-poc.yaml --threads 2 
    [2024-02-06 19:47:38] [INFO] [171386] [MainThread] [workflow_engine]: Starting workflow execution.
    [2024-02-06 19:47:38] [INFO] [171386] [MainThread] [workflow_engine]: Running tasks in parallel with 2 threads.
    [2024-02-06 19:47:38] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Initiating execution of task "allocate-manager".
    [2024-02-06 19:47:38] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Initiating execution of task "allocate-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:49:03] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Task "allocate-manager" completed succesfully in 84.95 seconds.
    [2024-02-06 19:49:03] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Initiating execution of task "provision-manager".
    [2024-02-06 19:49:11] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Task "allocate-linux-ubuntu-20.04-amd64" completed succesfully in 93.12 seconds.
    [2024-02-06 19:52:55] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Task "provision-manager" completed succesfully in 232.07 seconds.
    [2024-02-06 19:52:55] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Initiating execution of task "provision-install-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:53:31] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Task "provision-install-linux-ubuntu-20.04-amd64" completed succesfully in 35.93 seconds.
    [2024-02-06 19:53:31] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Initiating execution of task "run-agent-tests-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:53:31] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Initiating execution of task "provision-uninstall-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:53:51] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Task "provision-uninstall-linux-ubuntu-20.04-amd64" completed succesfully in 20.32 seconds.
    [2024-02-06 19:54:01] [INFO] [171386] [ThreadPoolExecutor-0_0] [workflow_engine]: Task "run-agent-tests-linux-ubuntu-20.04-amd64" completed succesfully in 30.15 seconds.
    [2024-02-06 19:54:01] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Initiating execution of task "run-agent-tests-uninstall-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:54:30] [INFO] [171386] [ThreadPoolExecutor-0_1] [workflow_engine]: Task "run-agent-tests-uninstall-linux-ubuntu-20.04-amd64" completed succesfully in 29.07 seconds.
    [2024-02-06 19:54:30] [INFO] [171386] [MainThread] [workflow_engine]: Executing cleanup tasks.
    [2024-02-06 19:54:30] [INFO] [171386] [ThreadPoolExecutor-1_0] [workflow_engine]: Initiating execution of task "allocate-linux-ubuntu-20.04-amd64".
    [2024-02-06 19:54:30] [INFO] [171386] [ThreadPoolExecutor-1_1] [workflow_engine]: Initiating execution of task "allocate-manager".
    [2024-02-06 19:54:38] [INFO] [171386] [ThreadPoolExecutor-1_1] [workflow_engine]: Task "allocate-manager" completed succesfully in 8.11 seconds.
    [2024-02-06 19:54:38] [INFO] [171386] [ThreadPoolExecutor-1_0] [workflow_engine]: Task "allocate-linux-ubuntu-20.04-amd64" completed succesfully in 8.33 seconds.
    ```
- **Execution with log level `DEBUG`**
    ```shellsession
    python workflow_engine.py /home/quebim/Wazuh/repos/wazuh-qa/deployability/modules/workflow_engine/examples/dtt1-agents-poc.yaml --log-level DEBUG
    [2024-02-06 20:23:20] [DEBUG] [180033] [MainThread] [workflow_engine]: Loading schema file: /home/quebim/Wazuh/repos/wazuh-qa/deployability/modules/workflow_engine/schemas/schema_v1.json
    [2024-02-06 20:23:20] [DEBUG] [180033] [MainThread] [workflow_engine]: Loading yaml file: /home/quebim/Wazuh/repos/wazuh-qa/deployability/modules/workflow_engine/examples/dtt1-agents-poc.yaml
    [2024-02-06 20:23:20] [INFO] [180033] [MainThread] [workflow_engine]: Starting workflow execution.
    [2024-02-06 20:23:20] [INFO] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Initiating execution of task "allocate-manager".
    [2024-02-06 20:23:20] [DEBUG] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Running task "allocate-manager" with arguments: ['allocation.py', '--action=create', '--provider=vagrant', '--size=large', '--composite-name=linux-ubuntu-20.04-amd64', '--inventory-output=/tmp/dtt1-poc/manager-linux-ubuntu-20.04-amd64/inventory.yaml', '--track-output=/tmp/dtt1-poc/manager-linux-ubuntu-20.04-amd64/track.yaml']
    [2024-02-06 20:24:15] [DEBUG] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Finished task "allocate-manager" execution with result:
    [2024-02-06 20:23:20] [INFO] ALLOCATOR: Creating instance at /tmp/wazuh-qa
    [2024-02-06 20:23:20] [DEBUG] ALLOCATOR: No config provided. Generating from payload
    [2024-02-06 20:23:24] [DEBUG] ALLOCATOR: Vagrantfile created. Creating instance.
    [2024-02-06 20:23:24] [INFO] ALLOCATOR: Instance VAGRANT-D28DA015-99D5-4A03-A34B-9BDCA7387E9E created.
    [2024-02-06 20:24:12] [INFO] ALLOCATOR: Instance VAGRANT-D28DA015-99D5-4A03-A34B-9BDCA7387E9E started.
    [2024-02-06 20:24:15] [DEBUG] ALLOCATOR: Using provided credentials
    [2024-02-06 20:24:15] [INFO] ALLOCATOR: Inventory file generated at /tmp/dtt1-poc/manager-linux-ubuntu-20.04-amd64/inventory.yaml
    [2024-02-06 20:24:15] [INFO] ALLOCATOR: Track file generated at /tmp/dtt1-poc/manager-linux-ubuntu-20.04-amd64/track.yaml
    
    [2024-02-06 20:24:15] [INFO] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Task "allocate-manager" completed succesfully in 54.88 seconds.
    [2024-02-06 20:24:15] [INFO] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Initiating execution of task "allocate-linux-ubuntu-20.04-amd64".
    [2024-02-06 20:24:15] [DEBUG] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Running task "allocate-linux-ubuntu-20.04-amd64" with arguments: ['allocation.py', '--action=create', '--provider=vagrant', '--size=small', '--composite-name=linux-ubuntu-20.04-amd64', '--inventory-output=/tmp/dtt1-poc/agent-linux-ubuntu-20.04-amd64/inventory.yaml', '--track-output=/tmp/dtt1-poc/agent-linux-ubuntu-20.04-amd64/track.yaml']
    [2024-02-06 20:25:17] [DEBUG] [180033] [ThreadPoolExecutor-0_0] [workflow_engine]: Finished task "allocate-linux-ubuntu-20.04-amd64" execution with result:
    [2024-02-06 20:24:15] [INFO] ALLOCATOR: Creating instance at /tmp/wazuh-qa
    [2024-02-06 20:24:15] [DEBUG] ALLOCATOR: No config provided. Generating from payload
    [2024-02-06 20:24:19] [DEBUG] ALLOCATOR: Vagrantfile created. Creating instance.
    [2024-02-06 20:24:19] [INFO] ALLOCATOR: Instance VAGRANT-F24F1108-D336-4481-9438-5088738DC4A1 created.
    [2024-02-06 20:25:13] [INFO] ALLOCATOR: Instance VAGRANT-F24F1108-D336-4481-9438-5088738DC4A1 started.
    [2024-02-06 20:25:17] [DEBUG] ALLOCATOR: Using provided credentials
    [2024-02-06 20:25:17] [INFO] ALLOCATOR: Inventory file generated at /tmp/dtt1-poc/agent-linux-ubuntu-20.04-amd64/inventory.yaml
    [2024-02-06 20:25:17] [INFO] ALLOCATOR: Track file generated at /tmp/dtt1-poc/agent-linux-ubuntu-20.04-amd64/track.yaml
    ```

### CODE REVIEW

**Review by pro-akim — APPROVED** [review]
## Review Notes
LGTM

File: deployability/modules/workflow_engine/logging/logger.py
Comment (fcaffieri) [inline deployability/modules/workflow_engine/logging/logger.py:1]: Remove unused comment lines

File: deployability/modules/workflow_engine/logging/logger.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import logging.config
+from pathlib import Path
+import threading
+
+import yaml
+
+# The logging module will use the generic logger anyway
+# that is because the logging module is a singleton.
+# So, we can just use it this way here, and it will work.
+# formatter = logging.Formatter('[%(asctime)s] [%(levelname)s] [%(process)d] [%(threadName)s] %(name)s: %(message)s')
```
Comment (QU3B1M) [inline deployability/modules/workflow_engine/logging/logger.py:11]: This should be removed
```suggestion
```

File: deployability/modules/workflow_engine/logging/logger.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#         log_level = 'DEBUG'
+
+#     if log_level.strip().upper() == 'DEBUG':
+#         logger.debug(log_message, extra=extra)
+#     elif log_level.strip().upper() == 'INFO':
+#         logger.info(log_message, extra=extra)
+#     elif log_level.strip().upper() == 'WARNING':
+#         logger.warning(log_message, extra=extra)
+#     elif log_level.strip().upper() == 'ERROR':
+#         logger.error(log_message, extra=extra)
```
Comment (QU3B1M) [inline deployability/modules/workflow_engine/logging/logger.py:69]: This should be also removed 
```suggestion
```

**Review by fcaffieri — CHANGES_REQUESTED** [review]

File: deployability/modules/workflow_engine/logging/logger.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    Loads the logging configuration from 'config.yaml' file.
+    """
+    config_path = Path(__file__).parent / 'config.yaml'
+    with open(config_path, 'r') as f:
+        config = yaml.safe_load(f.read())
+        logging.config.dictConfig(config)
+
+_load_config()
+
+    # self.logger = logging.getLogger(name)
```
Comment (QU3B1M) [inline deployability/modules/workflow_engine/logging/logger.py:24]: Remove
```suggestion
```

File: deployability/modules/workflow_engine/logging/logger.py
Comment (QU3B1M) [reply deployability/modules/workflow_engine/logging/logger.py:1]: Comments removed in last commit!

Comment (fcaffieri) [reply deployability/modules/workflow_engine/logging/logger.py:1]: [a60c1a5](https://github.com/wazuh/wazuh-qa/pull/4919/commits/a60c1a5551163c719a4224d190fea25701d295ae)

**Review by fcaffieri — APPROVED** [review]
LGTM

**Pull request merged**